### ISSUE TITLE
Feature Request: add option to use cross-validation to choosing split at each node

### ISSUE THREAD

**Issue by peterloleungyau** [dmlc/xgboost#9560]
There is a recent paper entitled "Feature Importance in Gradient Boosting Trees with Cross-Validation Feature Selection" (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9140774/) that attempt to address the well known problem of bias towards variables with large cardinalities in feature importance for tree models.

The idea is simple:
- at each node, use cross-validation to select the best split
- i.e. at each node, there are some samples, the split the samples into folds
    - for each fold, use the training portion to select metrics (e.g. impurity) as usual, but evaluate the effectiveness on the testing portion
    - then for each potential split, average the metrics across folds, then select the best one
    - if the average CV metrics is not better than no splitting, then stop splitting

Benefits of this method:

- the paper shows that this can alleviate the bias of feature importance, especially for more efficient feature importance such as gain importance. This is particularly important for real world problems where the number of features is large, and the sample size is also large, so we want to have efficient yet unbiased methods to calculate feature importance, to help feature selection.
- it also provides a natural way to stop growing the tree, i.e. if the CV metrics does not show improvement over not splitting. This should help with preventing overfitting, and also makes hyperparameter tuning easier.

The paper provides reference implementation for simple gradient boosting at https://github.com/aba27059/unbiased_fi_for_gb

It would be great if this can be added to XGBoost.

**Comment by rohan472000** [comment]
If we try to implement above then what cross-validation method should be used for it and which evaluation metric would be correct to measure the goodness of a split?

**Comment by peterloleungyau** [comment]
I think a simple K-fold cross-validation could be used, of course with a user configurable K. Not sure if user configurable stratified splitting would be useful.

And for the metric, I think we may use the same metric that XGBoost is optimizing, which I think should be the regularized loss (whichever loss function is used). This CVB is a general modification of choosing the feature to split, and it is applicable to any metric we are optimizing.

To my understanding of Section 3 of the paper https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9140774/#sec3-entropy-24-00687title

The CVB algorithm is something like this:

- at a node that we want to choose a feature for splitting:
    - first randomly assign the samples of the node (A) into K folds, call them A_1, A_2, ..., A_K
    - for each candidate feature v (after any column subsampling of features):
        - for k = 1 ... K:
            - use all A_i but A_k, i.e. A - A_k to choose a splitting point for feature v, as usual, optimizing the usual metric
            - having got the splitting point, use the chosen splitting point on A_k, and calculate value of the metric (that we are optimizing), call it L_{v,k}
        - average the metric over the folds, as a measure of goodness of choosing feature v for splitting, i.e. the goodness of v is L_v = (L_{v,1} + L_{v,2} + ... + L_{v,K})/K
    - choose the feature v' with the best L_v'
    - if L_v' is not better than the loss before splitting, then stop growing the tree at the current node
    - otherwise, use all the samples of the node (A) and v to choose a splitting point, as usual, and split the node accordingly

**Comment by trivialfis** [comment]
I haven't gone through the paper yet, but based on the descriptions here it seems not be not very efficient in terms of computation performance:
- We use histogram to approximate the gain function except for the `exact` tree method. This implies that during split evaluation, we don't actually go through the entire dataset but an approximation of it. Using the `exact` tree method is extremely slow in practice. I'm not sure what's the effect of splitting the histogram for evaluation.
- There are other more efficient ways to overfitting, from sampling to many regularization options to drop out. I haven't seen a comparison between these methods and would prefer methods that have reasonable computation performance and scalable to distributed systems.

**Comment by peterloleungyau** [comment]
The main aim of this CVB algorithm is to reduce the bias of feature importance (the default being gain importance) for variables with high cardinalities. That it also naturally gives a tree depth for each branch is a nice side effect.

I am not too familiar with the intricacies of the splitting in XGBoost, especially regarding the use of histogram.

- do you mean that a separate histogram is constructed from separate subsets of data (of the node) for each candidate feature, for splitting? or is the same subset of data used for all candidate features to construct separate histograms in a node?
- is the histogram used only for generating candidate split points for is it used for evaluating the "goodness" of split points as well?

Either way, with the use of histograms, I think two approaches are still worth considering:

1. still follow the basic flow of CVB (see the pseudo code in previous reply above):
    - at each node, random assign the samples to K folds
    - still for each feature and for each fold, use the training portion and the usual way to choose a split point
        - i.e. build histogram as usual, then choose a split point
    - but evaluate the split point on the testing point of the fold, as the "goodness" of splitting by that feature on that fold
    - then average the "goodness" across folds, as the "goodness" of splitting by a feature
    - then choose the feature for the usual split, which includes the usual construction of histogram
    - the computation performance penalty is roughly K times as usual, which is expected if the user, if the user chooses to turn on this CVB option.

2. change the flow of CVB a little bit:
    - do not use K fold, but rather follows train/validate split
    - at a node, for each feature
        - construct histogram as usual on a subset of data (of the node), to choose a split point
        - but evaluate the "goodness" of the split point on another subset of samples (of the node) that does not overlap with that set for constructing the histogram
        - this "goodness" becomes the "goodness" of the feature for splitting
        - once chosen the feature, then do the splitting using the chosen split point as usual
    - the computation performance penalty is smaller than using K-fold, but again, the user would also expect the higher computation cost if the user chooses to turn on this CVB.

The main idea of CVB is to estimate the generalization of choosing a certain feature for splitting at a give node, I think either of the above should still attain this goal, and can be adapted to the way the of choosing the split point, either using histogram or not.
